### ISSUE TITLE
Add wayoint's personal note data to exported GPX

### ISSUE THREAD

**Issue by baiti** [cgeo/cgeo#9234]
**Is your feature request related to a problem?**
Waypoints in cgeo support a free form text field, similar to the "Personal Cache Notes" at the geocache object level. Upon Export->GPX, the latter goes into the GPX file but the former does not.

**Describe the solution you'd like**
Add the personal notes data that is associated with a waypoint in cgeo to the exported GPX.

**Additional context**
Let me use this scenario to open a broader discussion: What is the intent of providing the capability to add custom waypoints to a geocache object in cgeo? The waypoint has similar properties than those that come down from GC.com but their sole purpose seems to be limited to using it on cgeo itself. There is no way back to geocaching.com (or any other such platform I believe).

In my workflow, a tool lilke cgeo is merely a locally cached copy of data that is authoritatively stored at geocaching.com. The cached data makes sense in situations where I don't have mobile data, out in the field on a geoaching.trip. Usually, if I make any changes out in the field. e.g. changing corrected coordinates, adding or changing personal cache notes, I always reflect that back to geocaching.com because that is the one and only place where it authoritatively belongs to. Why is that so relevant with the "autoritative source"? Like many others I use multiple different tools to access and manipulate data from geoaching.com, e.g. GSAK, Locus Map, cgeo and the list may go on. They all have in common: They depend on one single place where they fetch their data from and they all go to geocaching.com to get it from there.  Whatever such a tool modifies or adds, it must send this data back to geoaching.com where all other tools will "refresh" their local copy of the data in regular time intervals or as needed and triggered by me. Each and every other workflow will undoubtedly lead to different data on different devices and now way to ever get it back in sync again.

Back to the original topic: If I add a new waypoint in cgeo, there doesn't seem to be any way to reflect that back to geoaching.com. Is that a limitation in geocaching.com? At least their API supports user created waypoints:

Create / Update / Delete User Wayoint

To close the loop: Since there is no way to use the data in the personal notes field of a waypoint in a programmatic way, "Export -> GPX" seems to be the only way to get to it but only once the export function adds it to the GPX.

**Comment by MagpieFourtyTwo** [comment]
[custom waypoints]
> There is no way back to geocaching.com (or any other such platform I believe).

Yep. At least not as WPs, as even the GUI on gc.com does not allow to add WPs to a listing that's not yours (AFAIK).

> If I add a new waypoint in cgeo, there doesn't seem to be any way to reflect that back to geoaching.com.

Au contraire: With ``Save WPs to PN`` there is at least a workaround.  ;)

> Is that a limitation in geocaching.com? At least their API supports user created waypoints:

I guess this is for your own caches only, as a) gc.com's GUI does not show any entry point to add WPs to caches (which are not yours) and b) although the API description is a bit unclear for ``Create``, saying "Opted-out user will return a 403 Forbidden status", but for ``Update`` and ``Delete`` it's way more concrete, saying "Only owner may update the user waypoint".

And I guess there's no way around, as even GC little helper II (which I'm heavily using for years now) does not give the opportunity to add/change a WP - and this tool would *for sure* would make use of such a wonderful feature, if it only exists, so I guess there is no such function for real.

This is BTW one of the reasons, why we are heavily using the cache's PN - cause the stuff put in there is available on (nearly) all client devices (apart from Garmin, IIRC).

**Comment by baiti** [comment]
> 
> Au contraire: With `Save WPs to PN` there is at least a workaround. ;)

Yes indeed ... but .. .where is that function? I haven't found it yet.

> 
> > Is that a limitation in geocaching.com? At least their API supports user created waypoints:
> 
> I guess this is for your own caches only, as a) gc.com's GUI does not show any entry point to add WPs to caches (which are not yours) and b) although the API description is a bit unclear for `Create`, saying "Opted-out user will return a 403 Forbidden status", but for `Update` and `Delete` it's way more concrete, saying "Only owner may update the user waypoint".
> 

Owner in what exact context? Owner of the geocache where the WP is associated to? Or owner in the sense of user who created the WP?

They treat their own "corrected coords" field somehow like a waypoint. And every user (premium) can add his/her own corrected coords to any listing that supports it. The functions: Upsert / Delete Coorected Coordinates is documented in the section that addrresses User Waypoint.

Well ... never mind, it is OT here.

> This is BTW one of the reasons, why we are heavily using the cache's PN - cause the stuff put in there is available on (nearly) all client devices (apart from Garmin, IIRC).

Yes, I am also heavily using it. Also for instance for documenting solutions for mysterys. The 2,500 char limit is a problem though but again, that is OT here ;)

Thanks for you clarifications ... and I urgently need that "Save WP to PN" feature

**Comment by eddiemuc** [comment]
> Thanks for you clarifications ... and I urgently need that "Save WP to PN" feature

You find it directly under the PN field. Just click the button circled below for a cache where you defined one or more waypoints.

![image](https://user-images.githubusercontent.com/6909759/96612948-adb14080-12fe-11eb-9bab-42cd5669c631.png)

**Comment by murggel** [comment]
> **Describe the solution you'd like**
> Add the personal notes data that is associated with a waypoint in cgeo to the exported GPX.

With #6914 the user note of a waypoint is added for gpx-export to the "normal" waypoint note, so the information is contained in the gpx

**Comment by baiti** [comment]
not sure what a "normal waypoint note" is. I understand and differntiate:

(1) user note in a waypoint of a geocache
(2) personal note (cgeo term) in the geocache object (actually referenced as "personal cache note" by gc.com)

geocaching.com doesn't seem to have a notion of (1) so it seems like we're merging all (1) into (2)? The result would go up to gc.com as well as into exported GPXes?

Is that what we're doing?

Do you by chance know whether #6914 is in release 2020.10.10? I have not seen that functionality in that release

**Comment by murggel** [comment]
> (1) user note in a waypoint of a geocache
> (2) personal note (cgeo term) in the geocache object (actually referenced as "personal cache note" by gc.com)
> 
> geocaching.com doesn't seem to have a notion of (1) so it seems like we're merging all (1) into (2)? The result would go up to gc.com as well as into exported GPXes?
> 
> Is that what we're doing?
> 
> Do you by chance know whether #6914 is in release 2020.10.10? I have not seen that functionality in that release



> not sure what a "normal waypoint note" is. I understand and differntiate:
> 
> (1) user note in a waypoint of a geocache
> (2) personal note (cgeo term) in the geocache object (actually referenced as "personal cache note" by gc.com)
> 
> geocaching.com doesn't seem to have a notion of (1) so it seems like we're merging all (1) into (2)? 
Correct
> The result would go up to gc.com as well as into exported GPXes?
> 

No, only to gpx, AFAIK we don't update waypoints.

> Do you by chance know whether #6914 is in release 2020.10.10? I have not seen that functionality in that release

No, it probably will be in next release

**Comment by baiti** [comment]
> > not sure what a "normal waypoint note" is. I understand and differntiate:
> > (1) user note in a waypoint of a geocache
> > (2) personal note (cgeo term) in the geocache object (actually referenced as "personal cache note" by gc.com)
> > geocaching.com doesn't seem to have a notion of (1) so it seems like we're merging all (1) into (2)?
> 
> Correct
> 
> > The result would go up to gc.com as well as into exported GPXes?
> 
> No, only to gpx, AFAIK we don't update waypoints.

Yes, that is a discussion we just had here: [#9233 ](https://github.com/cgeo/cgeo/issues/#9233)
At least gc.com doesn't (yet) have a notion of user defined waypoints although their API shows that they're probably working on it. And because they don't support it, it can only go into (a) personal cache notes and (b) the exported GPX.

> 
> > Do you by chance know whether #6914 is in release 2020.10.10? I have not seen that functionality in that release
> 
> No, it probably will be in next release

I am looking forward to testing it then.

Thanks much for clarifying the subject.

**Comment by MagpieFourtyTwo** [comment]
@baiti 
> Owner in what exact context?

Cache Owner

**Comment by Billy-Wallace** [comment]
For a custom waypoint that has been configured to solve multicaches with the **waypoint calculator** using the **Plain** option, I can use c:geo to successfully write to the Personal Notes and upload them to GC.com, for example:

```
{c:geo-start}
@Formula (F) (NO-COORD)
"N50 57.(A+E)(B+B)(E) W001 21.(D)(B+D)(A+B+B)
telephone number 
023806ABCDE
short walk
Checksum 30"
{c:geo-end}
```

Which is fantastic but when using c:geo on another device and downloading this cache, my expectation was that the same custom waypoint would be recreated ready for action in the field.

Unfortunately, the read process fails to populate this waypoint. I am guessing the reason is because the coordinates are not valid, that is, they are in the format of a valid formula format for the waypoint latitude and longitude but not one of the excepted valid coordinate formats:

```
Plain
N50 57.(A+E)(B+B)(E)
W001 21.(D)(B+D)(A+B+B)
A=|a+b| D=|a+b|
B=|a+b| E=|a+b|
Note ...
```

I see this style of waypoint calculations to be the most common usage, so it will be a shame if this cannot be automated to create this waypoint on load rather than going through copy/paste manual configurations, which is doable ... but a pain ...

If this was fixed, we could also do are planning in the cache page, updating the Personal Notes in the format template:

```
{c:geo-start}
@Title (F) (NO-COORD)
"Valid Lat/Long Formula for calculator
additional notes"
{c:geo-end}
```

Or in whatever format that works best for the WPC author.

Then, we load our GPX into c:geo and are good to go.

**Comment by murggel** [comment]
That would be a great feature for me as well, even, if the value of the variables are somehow exported too. I would use it in addition as archiv-functionaltiy in GSAK.
Or preparing there, calculate some values offline at home and the remaining out in the field. 
One more point, but this would be a separate issue:
Some multi have only one set of variables for all stages. So it would be great, if the value of the variables are populated to all waypoints mit formulas, probably only on demand or as an option...

**Comment by murggel** [comment]
> I am looking forward to testing it then.

Thanks 

> Thanks much for clarifying the subject.

:-)

**Comment by murggel** [comment]
Unfortunately I linked the commits to this issue instead to #9338.
Probably somebody can help me to fix this?

**Comment by murggel** [comment]
@baiti: can this issue be closed with #6914 or do you want to leave it open for the discussion at the start of the issue?
For the issues regarding the waypoint calculator i think there are separate issues...

**Comment by baiti** [comment]
I have verified the functionality that  copies waypoint data into the PCN field of the cache and exports that field to geocaching.com. I have not yet verified what happens if I export a GPX. I assume the waypoint data ends up in the PCNs there as well. Please give me a abit more time to verify that on 2020.11.03.

**Comment by murggel** [comment]
> Please give me a abit more time to verify that on 2020.11.03.
No problem. But it went not into 2020.11.03, since that was a bug fix release. 
You have to wait for the next release from master.
But I don't know how to distinguish ...

**Comment by baiti** [comment]
I only have one device at hand and I treat that as "production", so I only install official releases from play store. We will figure it out, it isn't really a serious issue.
